File: pubsub.py
from pika import BlockingConnection, ConnectionParameters

EXCHANGE_NAME = 'alexandra'

def create_channel():
    connection = BlockingConnection(ConnectionParameters('localhost'))
    channel = connection.channel()
    channel.exchange_declare(exchange=EXCHANGE_NAME, type='topic')
    return channel

def publish(channel, label, message):
    channel.basic_publish(exchange=EXCHANGE_NAME,
                          routing_key=label,
                          body=message)

def subscribe(channel, label):
    result = channel.queue_declare(exclusive=True)
    queue = result.method.queue
    channel.queue_bind(exchange=EXCHANGE_NAME,
                       queue=queue,
                       routing_key=label)
    return queue

def unsubscribe(channel, queue, label):
    channel.queue_unbind(exchange=EXCHANGE_NAME, queue=queue, routing_key=label)

def get_message(channel, queue):
    return channel.basic_get(0, queue=queue, no_ack=True)[2]

File: emcee/libraries/pubsub.rb
require "bunny"

module Pubsub

  class AMQPConnection
    def initialize(url, exchange_name)
      host = %r{amqp://(\w+)}.match(url)[1]
      conn = Bunny.new(:hostname => host)
      conn.start
      @channel = conn.create_channel
      @exchange = @channel.topic(exchange_name)
    end

    def publish(topic, message)
      @exchange.publish(message, :routing_key => topic)
    end

    def subscribe(topic)
      queue = @channel.queue("")
      queue.bind(@exchange, :routing_key => topic)
      return queue
    end

    def unsubscribe(queue)
      queue.delete
    end

    def consume(queue, &block)
      queue.subscribe(:block => true) do |di, p, body|
        block.call(body)
      end
    end

    def get_message(queue)
      delivery_info, properties, payload = queue.pop
      return payload
    end
  end

end
